File: 4_Dicas/4.5_Code_profiling/Ex4.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>

#define N 10000000

struct vetor_int
{
	int *x;
	int tam;
	double media;
};

void preencher_vetores(
	struct vetor_int *vx,
	struct vetor_int *vy);
void calc_media(
	struct vetor_int *v);
double distancia_euclidiana(
	struct vetor_int *v1,
	struct vetor_int *v2);

int main(int argc, char *argv[])
{
	struct vetor_int v1, v2;
	double dist_eucl;

	v1.x = (int *)malloc(N*sizeof(int));
	v1.tam = N;
	v2.x = (int *)malloc(N*sizeof(int));
	v2.tam = N;

	preencher_vetores(&v1, &v2);

	calc_media(&v1);
	calc_media(&v2);
	dist_eucl = distancia_euclidiana(&v1, &v2);

	printf("Resultado %s:\n", argv[0]);
	printf("   Media(x) = %1.10f\n",
		v1.media);
	printf("   Media(y) = %1.10f\n",
		v2.media);
	printf("   Dist_euclidiana(x,y) = %3.10f\n\n",
		dist_eucl);

	free(v1.x);
	free(v2.x);
	return 0;
}

void preencher_vetores(struct vetor_int *vx, struct vetor_int *vy)
{
	int i;
	for(i=0; i<(vx->tam); i++)
	{
		vx->x[i] = 1000/(i+1);
		vy->x[i] = 1000/(3*i*i+5*i+1);
	}
}

void calc_media(struct vetor_int *v)
{
	int i;
	for(v->media=0.0, i=0; i<(v->tam); i++)
		v->media += (double)v->x[i];
	v->media /= (double)v->tam;
}

double distancia_euclidiana(struct vetor_int *v1, struct vetor_int *v2)
{
	double dist_eucl, dummy;
	int i;

	for(dist_eucl=0.0, i=0; i<v1->tam; i++)
	{
		dummy = (double)v1->x[i];
		dummy -= (double)v2->x[i];
		dist_eucl += dummy*dummy;
	}
	return sqrt(dist_eucl);
}

File: 4_Dicas/4.5_Code_profiling/Octave/Ex4.m
N = 1000000;

function [a,b] = preencher_vetores(N)
	k = 0:(N-1);
	a = floor(1000./bitand(k+1, (2^32)-1));
	b = floor(1000./bitand(3*(k.*k)+5*k+1,(2^32)-1));
endfunction

function media_a = calc_media(a)
	media_a = mean(a);
endfunction

function dist_eucl = distancia_euclidiana(a, b)
	dist_eucl = sqrt(sum((a-b).^2));
endfunction

if length(argv())>1
	profile on
end
[x,y] = preencher_vetores(N);
media_x = calc_media(x);
media_y = calc_media(y);
dist_eucl = distancia_euclidiana(x,y);
if length(argv())>1
	profile off
	profshow()
end

disp("Resultado:")
disp(["   Media(x) = " num2str(media_x)]);
disp(["   Media(y) = " num2str(media_y)]);
disp(["   Dist_euclidiana(x,y) = " num2str(dist_eucl)]);